test/testdata/infer/generics/variance_methods.rb

```diff
@@ -22,24 +22,30 @@ def lookup_blk(key, &blk); end
   # should fail: the contravariant type key is returned
   sig {abstract.returns(In)}
   def returns_contravariant; end
+# ^^^^^^^^^^^^^^^^^^^^^^^^^ error: `type_member` `In` was defined as `:in` but is used in an `:out` context
 
   # should fail: the contravariant type key is returned
   sig {abstract.returns(T.nilable(In))}
   def returns_nilable_contravariant; end
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error: `type_member` `In` was defined as `:in` but is used in an `:out` context
 
   # should fail: the contravariant type key is given to the block as an argument
   sig {abstract.params(blk: T.proc.params(arg: In).void).void}
+                     # ^^^ error: `type_member` `In` was defined as `:in` but is used in an `:out` context
```

Comment:
_might_ be nice to keep track of some sort of nesting depth / number of times we negated? 

I think variance errors are probably always going to be confusing.

ooc, did you have a look at the error messages that other languages emit? curious how our error messages compare / if we can improve on them.

Reply:
I experimented with C# a bit. Its errors are similarly confusing, and convey about as much information as this one does now, but I'll try to stress it a bit tomorrow and see how bad I can make it.

Reply:
> I experimented with C# a bit.

I'm unsure if their errors are good examples to follow.
We could potentially list variances of all wrapper types and explain it, but let's see it become a problem and let's work on it separately.